addons/binding/org.openhab.binding.connectsdk/src/main/java/org/openhab/binding/connectsdk/handler/AbstractChannelHandler.java

```diff
@@ -0,0 +1,107 @@
+ *
+ * @author Sebastian Prehn
+ * @since 1.8.0
+ */
+public abstract class AbstractChannelHandler<T> implements ChannelHandler {
+    private static final Logger logger = LoggerFactory.getLogger(AbstractChannelHandler.class);
+
+    // IP to Subscriptions map
+    private Map<String, ServiceSubscription<T>> subscriptions;
+
+    // lazy init
+    private synchronized Map<String, ServiceSubscription<T>> getSubscriptions() {
+        if (subscriptions == null) {
+            subscriptions = new ConcurrentHashMap<String, ServiceSubscription<T>>();
+        }
+        return subscriptions;
+    }
+
+    @Override
+    public void onDeviceReady(final ConnectableDevice device, final String channelId, final ConnectSDKHandler handler) {
+        // NOP
+    }
+
+    @Override
+    public void onDeviceRemoved(final ConnectableDevice device, final String channelId,
+            final ConnectSDKHandler handler) {
+        // NOP
+    }
+
+    @Override
+    public final synchronized void refreshSubscription(final ConnectableDevice device, final String channelId,
+            final ConnectSDKHandler handler) {
+        removeAnySubscription(device);
+        if (handler.isChannelInUse(channelId)) { // only listen if least one item is configured
+            ServiceSubscription<T> listener = getSubscription(device, channelId, handler);
+            if (listener != null) {
+                logger.debug("Subscribed {} on IP: {}", this.getClass().getName(), device.getIpAddress());
+                getSubscriptions().put(device.getIpAddress(), listener);
+            }
+        }
+    }
+
+    /**
+     * Creates a subscription instance for this device. This may return <code>null</code> if no subscription is possible
+     * or required.
+     *
+     * @param device device to which state changes to subscribe to
+     * @param channelID channel ID
+     * @param handler
+     * @return instance or <code>null</code> if no subscription is possible or required
+     */
+    protected ServiceSubscription<T> getSubscription(final ConnectableDevice device, final String channelId,
+            ConnectSDKHandler handler) {
+        return null;
+    }
+
+    @Override
+    public final synchronized void removeAnySubscription(final ConnectableDevice device) {
+        if (subscriptions != null) { // only if subscriptions was initialized (lazy loading)
+            ServiceSubscription<T> l = subscriptions.remove(device.getIpAddress());
+            if (l != null) {
+                l.unsubscribe();
+                logger.debug("Unsubscribed {} on IP: {}", this.getClass().getName(), device.getIpAddress());
+            }
+        }
+    }
+
+    protected <O> ResponseListener<O> createDefaultResponseListener() {
+        return new ResponseListener<O>() {
+
+            @Override
+            public void onError(ServiceCommandError error) {
+                logger.error("{}: {}.", this.getClass().getName(), error.getMessage());
```

Comment:
Don't log to error. error logging is reserved for non-recoverable system errors in the openHAB system.

Reply:
Changed them all to warn level.